agent/lib/kontena/rpc/service_pods_api.rb

```diff
@@ -4,18 +4,36 @@ module Kontena
   module Rpc
     class ServicePodsApi
 
-      # @param [String] service_id
-      # @param [Integer] instance_number
+      # @param service_id [String]
+      # @param instance_number [Integer]
       # @return [Hash]
       def restart(service_id, instance_number)
-        Kontena::ServicePods::Restarter.new(service_id, instance_number).perform
+        service_pod = find_service_pod(service_id, instance_number)
+        raise Kontena::RpcServer::Error.new(404, "Service instance not found") unless service_pod
+
+        Kontena::ServicePods::Restarter.new(service_pod).perform
         {}
       end
 
       def notify_update(reason)
         Celluloid::Notifications.publish('service_pod:update', reason)
         {}
       end
+
+      private
+
+      # @param service_id [String]
+      # @param instance_number [Integer]
+      def find_service_pod(service_id, instance_number)
+        worker = Celluloid::Actor[:service_pod_manager].workers.find { |w|
+          w.service_pod.service_id == service_id && w.service_pod.instance_number == instance_number
```

Comment:
Hmm, feels like a bad idea to share the `Hash` across actor threads... this should really be something like `Celluloid::Actor[:service_pod_manager].lookup_service_pod(service_id, instance_number)`

Reply:
The two calls to `w.service_pod` are both also sync actor calls to each `ServicePodWorker`. If those actors are blocked in the exclusive `apply`, then this can also end up blocking the RPC server for a long time.

Reply:
Yes, this was a bad idea. Reverted these changes.